Lib/glyphsLib/builder/kerning.py

```diff
@@ -20,31 +20,47 @@
 
 def to_ufo_kerning(self):
     for master in self.font.masters:
+        master_id = master.id
         kerning_source = master.metricsSource  # Maybe be a linked master
         if kerning_source is None:
             kerning_source = master
-        if kerning_source.id in self.font.kerning:
-            kerning = self.font.kerning[kerning_source.id]
-            _to_ufo_kerning(self, self._sources[master.id].font, kerning)
+        if kerning_source.id in self.font.kerningLTR:
+            kerning = self.font.kerningLTR[kerning_source.id]
+            _to_ufo_kerning(self, self._sources[master_id].font, kerning)
+        if kerning_source.id in self.font.kerningRTL:
+            kerning = self.font.kerningRTL[kerning_source.id]
+            _to_ufo_kerning(self, self._sources[master_id].font, kerning, "RTL")
 
 
-def _to_ufo_kerning(self, ufo, kerning_data):
+def _to_ufo_kerning(self, ufo, kerning_data, direction="LTR"):
     """Add .glyphs kerning to an UFO."""
 
     warning_msg = "Non-existent glyph class %s found in kerning rules."
 
     for left, pairs in kerning_data.items():
-        match = re.match(r"@MMK_L_(.+)", left)
+        if direction == "LTR":
+            match = re.match(r"@MMK_L_(.+)", left)
+        elif direction == "RTL":
+            match = re.match(r"@MMK_R_(.+)", left)
         left_is_class = bool(match)
         if left_is_class:
-            left = "public.kern1.%s" % match.group(1)
+            if direction == "LTR":
+                left = "public.kern1.%s" % match.group(1)
+            elif direction == "RTL":
+                left = "public.kern1.%s.RTL" % match.group(1)
             if left not in ufo.groups:
                 self.logger.warning(warning_msg % left)
         for right, kerning_val in pairs.items():
-            match = re.match(r"@MMK_R_(.+)", right)
+            if direction == "LTR":
+                match = re.match(r"@MMK_R_(.+)", right)
+            elif direction == "RTL":
+                match = re.match(r"@MMK_L_(.+)", right)
             right_is_class = bool(match)
             if right_is_class:
-                right = "public.kern2.%s" % match.group(1)
+                if direction == "LTR":
+                    right = "public.kern2.%s" % match.group(1)
+                elif direction == "RTL":
+                    right = "public.kern2.%s.RTL" % match.group(1)
```

Comment:
this repetition of if/elif blocks with minor differences feels a bit verbose, if we could factor some of this logic out and reuse would look nicer

Reply:
also, watch out for if statements without an else branch, if `direction` is neither LTR nor RTL (I know, should be impossible) then we would incur in a NameError for the `match` local variable isn't initialized. Maybe do `else:` and then assert that it's the other value that you expect it to be so we enver fall off the if

Reply:
I can offer this:

```python
def _to_ufo_kerning(self, ufo, kerning_data, direction="LTR"):
    """Add .glyphs kerning to an UFO."""

    class_missing_msg = "Non-existent glyph class %s found in kerning rules."
    overwriting_kerning_msg = "Overwriting kerning value for %s."

    for left, pairs in kerning_data.items():
        match = re.match(
            r"@MMK_" + ("L" if direction == "LTR" else "R") + "_(.+)", left
        )
        left_is_class = bool(match)
        if left_is_class:
            left = "public.kern1.%s" % match.group(1)
            if direction == "RTL":
                left += ".RTL"
            if left not in ufo.groups:
                self.logger.warning(class_missing_msg % left)
        for right, kerning_val in pairs.items():
            match = re.match(
                r"@MMK_" + ("R" if direction == "LTR" else "L") + "_(.+)", right
            )
            right_is_class = bool(match)
            if right_is_class:
                right = "public.kern2.%s" % match.group(1)
                if direction == "RTL":
                    right += ".RTL"
                if right not in ufo.groups:
                    self.logger.warning(class_missing_msg % right)

            if left in ufo.kerning and right in ufo.kerning[left]:
                self.logger.warning(overwriting_kerning_msg % [left, right])
            ufo.kerning[left, right] = kerning_val
```

Reply:
it is an improvement over the first iteration of the code, but I was thinking more in terms of reusing logic by parameterizing functions e.g.

```python
def _matches_mmk_class(name: str, direction: str, side: int) -> Optional[re.Match]:
    ...
```

or

```python
def _ufo_class_name(name, direction, side) -> str
```

you got the hang of it

Reply:
How about this? It became more code overall, tho:

```python
def _to_ufo_kerning(self, ufo, kerning_data, direction="LTR"):
    """Add .glyphs kerning to an UFO."""

    class_missing_msg = "Non-existent glyph class %s found in kerning rules."
    overwriting_kerning_msg = "Overwriting kerning value for %s."

    for left, pairs in kerning_data.items():
        left, left_is_class = _ufo_class_name(left, direction, "left")
        if left_is_class and left not in ufo.groups:
            self.logger.warning(class_missing_msg % left)

        for right, kerning_val in pairs.items():
            right, right_is_class = _ufo_class_name(right, direction, "right")
            if right_is_class and right not in ufo.groups:
                self.logger.warning(class_missing_msg % right)

            # Emit kerning
            if left in ufo.kerning and right in ufo.kerning[left]:
                self.logger.warning(overwriting_kerning_msg % [left, right])
            ufo.kerning[left, right] = kerning_val


def _ufo_class_name(name, direction, side):
    """Return the UFO class name for a .glyphs class name."""
    if side == "left":
        side_key = "L" if direction == "LTR" else "R"
        side_number = "1"
    else:
        side_key = "R" if direction == "LTR" else "L"
        side_number = "2"

    match = re.match(r"@MMK_" + side_key + r"_(.+)", name)

    name_is_class = bool(match)
    if name_is_class:
        name = "public.kern%s.%s" % (side_number, match.group(1))
        if direction == "RTL":
            name += ".RTL"

    return name, name_is_class
```

Reply:
it's looking better, but don;t use the terms left / right when you actually mean logical 1 or 2 position. Maybe "side" is not the right word (as it's tied to actual left/right), how about "order" (1 or 2)


(untested)

```python
import re


def _to_ufo_kerning(self, ufo, kerning_data, direction="LTR"):
    """Add .glyphs kerning to an UFO."""

    class_missing_msg = "Non-existent glyph class %s found in kerning rules."
    overwriting_kerning_msg = "Overwriting kerning value for %s."

    for first, pairs in kerning_data.items():
        first, is_class = _ufo_class_name(first, direction, 1)
        if is_class and first not in ufo.groups:
            self.logger.warning(class_missing_msg, first)

        for second, kerning_val in pairs.items():
            second, is_class = _ufo_class_name(second, direction, 2)
            if is_class and second not in ufo.groups:
                self.logger.warning(class_missing_msg, second)

            if (first, second) in ufo.kerning:
                self.logger.warning(overwriting_kerning_msg, first, second)
            ufo.kerning[first, second] = kerning_val


def _ufo_class_name(name, direction, order):
    """Return the UFO class name for a .glyphs class name."""
    if order == 1:
        side = "L" if direction == "LTR" else "R"
    else:
        assert order == 2
        side = "R" if direction == "LTR" else "L"

    match = re.match(fr"@MMK_{side}_(.+)", name)

    name_is_class = bool(match)
    if name_is_class:
        name = f"public.kerns{order}.{match.group(1)}"
        if direction == "RTL":
            name += ".RTL"

    return name, name_is_class
```

Reply:
The words left and right were already there. I want to fix the logic first and then rename the variables.

Reply:
they only make things more confusing, even more so now that we added the RTL logic, what they mean is actually logical first and second, since you're touching this code, I'd recommend you also do the renaming for clarity's sake

Reply:
Anyway, there was one little mistake in your code, otherwise I took it as is.